src/emulator/functionsRuntimeWorker.ts

```diff
@@ -0,0 +1,210 @@
+    this.runtime.send(args);
+  }
+
+  get state(): RuntimeWorkerState {
+    return this._state;
+  }
+
+  set state(state: RuntimeWorkerState) {
+    if (state === RuntimeWorkerState.IDLE) {
+      // Remove all temporary log listeners every time we move to IDLE
+      for (const l of this.logListeners) {
+        this.runtime.events.removeListener("log", l);
+      }
+      this.logListeners = [];
+    }
+
+    if (state === RuntimeWorkerState.FINISHED) {
+      this.runtime.events.removeAllListeners();
+    }
+
+    this.log(state);
+    this._state = state;
+    this.stateEvents.emit(this._state);
+  }
+
+  onLogs(listener: LogListener, forever: boolean = false) {
+    if (!forever) {
+      this.logListeners.push(listener);
+    }
+
+    this.runtime.events.on("log", listener);
+  }
+
+  waitForDone(): Promise<any> {
+    if (this.state === RuntimeWorkerState.IDLE || this.state === RuntimeWorkerState.FINISHED) {
+      return Promise.resolve();
+    }
+
+    return new Promise((res) => {
+      const listener = () => {
+        this.stateEvents.removeListener(RuntimeWorkerState.IDLE, listener);
+        this.stateEvents.removeListener(RuntimeWorkerState.FINISHED, listener);
+        res();
+      };
+
+      // Finish on either IDLE or FINISHED states
+      this.stateEvents.once(RuntimeWorkerState.IDLE, listener);
+      this.stateEvents.once(RuntimeWorkerState.FINISHED, listener);
+    });
+  }
+
+  waitForSystemLog(filter: (el: EmulatorLog) => boolean): Promise<EmulatorLog> {
+    return EmulatorLog.waitForLog(this.runtime.events, "SYSTEM", "runtime-status", filter);
+  }
+
+  private log(msg: string): void {
+    EmulatorLogger.log("DEBUG", `[worker-${this.triggerId}-${this.id}]: ${msg}`);
+  }
+}
+
+export class RuntimeWorkerPool {
+  workers: { [triggerId: string]: Array<RuntimeWorker> } = {};
+
+  /**
+   * When code changes (or in some other rare circumstances) we need to get
+   * a new pool of workers. For each IDLE worker we kill it immediately. For
+   * each BUSY worker we move it to the FINISHING state so that it will
+   * kill itself after the next run.
+   */
+  refresh() {
+    Object.values(this.workers).forEach((arr) => {
+      arr.forEach((w) => {
+        if (w.state === RuntimeWorkerState.IDLE) {
+          this.log(`Shutting down IDLE worker (${w.triggerId})`);
+          w.runtime.shutdown();
+        } else if (w.state === RuntimeWorkerState.BUSY) {
+          this.log(`Marking BUSY worker to finish (${w.triggerId})`);
+          w.state = RuntimeWorkerState.FINISHING;
+        }
+      });
+    });
+  }
+
+  /**
+   * Immediately kill all workers.
+   */
+  exit() {
+    Object.values(this.workers).forEach((arr) => {
+      arr.forEach((w) => {
+        if (w.state === RuntimeWorkerState.IDLE) {
+          w.runtime.shutdown();
+        } else {
+          w.runtime.kill();
+        }
+      });
+    });
+  }
+
+  getIdleWorker(triggerId: string | undefined): RuntimeWorker | undefined {
```

Comment:
why does this accept `undefined`? I think I can see the use of returning undefined, but I don't see any use case in your code of accepting `undefined`...

Reply:
`FunctionsRuntimeBundle` has `triggerId?: string`.  A bundle with no `triggerId` means "I want to run the runtime but no function in particular.  It's used to diagnose the user environment.